### PR TITLE
Move proposal across profiles. Fixes #216

### LINKED ISSUE #216: Ability to move a proposal to a proposal space under a different profile
Current, funnel lets a proposal manager to transfer proposals from one proposal space to another, under the same profile. While this is useful, there's a requirement for moving a proposal to a proposal under a different profile too, given the proposal manager is allowed to manage proposals for the other profile too.

### PR DESCRIPTION
**Pull request by iambibhas**
Fixes #216.

This is how it is in production right now 

selection_223

This is how it is after the changes introduced 

selection_222

My user is in admin team of metarefresh 2017. So looks like, currently, if you are in admin team of Metarefresh 2017, you can move proposals in any other proposal space in metarefresh profile even though you are not in admin_team of any of them, nor in admin team of metarefresh. Is that ok?

My proposed logic is, you can only move proposals to spaces you are admin of, or all the spaces within the profiles you are admin of. Does that sound right?

### CODE REVIEW

File: funnel/views/proposal.py [hunk 1]
```diff
@@ -398,9 +406,15 @@ def proposal_moveto(profile, space, proposal):
...
+
+    if not target_space:
         abort(404)
+
+    if not (target_space.profile.admin_team in g.user.teams or target_space.admin_team in g.user.teams):
+        flash(_("You do not have permission to move this proposal to {space}.".format(space=target_space.title)))
+        abort(403)
+
+    proposal.move_to(target_space)
+    db.session.commit()  # WARNING: GET request!
```
Comment (shreyas-satish) [inline funnel/views/proposal.py]: Is this still a GET request?

Comment (iambibhas) [reply funnel/views/proposal.py]: Yes. I was just think about making it a PUT or POST request.

**Comment by iambibhas** [comment]
selection_224

File: funnel/forms/proposal.py [hunk 2]
```diff
@@ -77,3 +77,7 @@ class ProposalStatusForm(forms.Form):
     status = forms.SelectField(
         __("Status"), coerce=int,
         choices=[(status, label.title) for (status, label) in PROPOSALSTATUS.items() if status != PROPOSALSTATUS.DRAFT])
+
+
+class ProposalMoveForm(forms.Form):
+    target = forms.SelectField(__("Move Proposal To"))
```
Comment (shreyas-satish) [inline funnel/forms/proposal.py]: Sentence case.

File: funnel/templates/proposal.html.jinja2 [hunk 3]
```diff
@@ -133,20 +133,11 @@
             {% endif %}
           {% endif %}
           {% if proposal.session %}
-            <br><a class="btn btn-xs btn-info" href="{{ proposal.session.url_for() }}">View session in schedule</a>
+            <a class="btn btn-xs btn-info" href="{{ proposal.session.url_for() }}">View session in schedule</a>
           {% endif %}
-          {%- if 'move-proposal' in g.permissions %}
+          {%- if 'move-proposal' in g.permissions and proposal_move_form %}
```
Comment (shreyas-satish) [inline funnel/templates/proposal.html.jinja2]: is `and proposal_move_form` required?

File: funnel/views/proposal.py [hunk 4]
```diff
@@ -299,10 +299,22 @@ def proposal_view(profile, space, proposal):
...
+    proposal_move_form = None
+    if 'move-proposal' in space.permissions(g.user, g.permissions):
+        team_ids = [t.id for t in g.user.teams]
+        movable_spaces = ProposalSpace.query.join(ProposalSpace.profile).filter(
+            (ProposalSpace.admin_team_id.in_(team_ids)) |
+            (Profile.admin_team_id.in_(team_ids))
+            ).order_by(ProposalSpace.date.desc())
+
+        proposal_move_form = ProposalMoveForm()
+        proposal_move_form.target.choices = [(space.id, space.title) for space in movable_spaces]
```
Comment (shreyas-satish) [inline funnel/views/proposal.py]: Use `set_queries` inside the form to set choices. See https://github.com/hasgeek/boxoffice/blob/master/boxoffice/forms/discount.py#L50 for an example.

File: funnel/views/proposal.py [hunk 5]
```diff
@@ -386,21 +398,30 @@ def proposal_prev(profile, space, proposal):
...
     ((Proposal, ProposalRedirect), {'url_name': 'proposal', 'proposal_space': 'space'}, 'proposal'),
     permission='move-proposal', addlperms=lastuser.permissions)
 def proposal_moveto(profile, space, proposal):
-    target_name = request.args.get('target')
-    if not target_name:
+    csrf_form = Form()
+    if not csrf_form.validate_on_submit():
+        abort(403)
+
+    target_space_id = request.values.get('target')
```
Comment (shreyas-satish) [inline funnel/views/proposal.py]: Read this value from the `form` object.

File: funnel/templates/proposal.html.jinja2 (hunk 3 above)
Comment (iambibhas) [reply funnel/templates/proposal.html.jinja2]: Not really. I'll remove it.

File: funnel/views/proposal.py (hunk 4 above)
Comment (iambibhas) [reply funnel/views/proposal.py]: @shreyas-satish that's for `QuerySelectMultipleField`, this is a basic `SelectField`.

File: funnel/views/proposal.py [hunk 6]
```diff
@@ -386,21 +390,28 @@ def proposal_prev(profile, space, proposal):
...
-    target_name = request.args.get('target')
-    if not target_name:
-        abort(404)
-    profile_name, space_name = target_name.split('/', 1)
-    target_space = ProposalSpace.query.filter(ProposalSpace.name == space_name).join(Profile).filter(Profile.name == profile_name).one_or_none()
-    if target_space:
-        proposal.move_to(target_space)
-        db.session.commit()  # WARNING: GET request!
+    csrf_form = Form()
+    if not csrf_form.validate_on_submit():
```
Comment (shreyas-satish) [inline funnel/views/proposal.py]: Why is `csrf_form` required? Shouldn't `proposal_move_form` suffice?

Comment (iambibhas) [reply funnel/views/proposal.py]: Oh yeah, that form itself contains csrf token.

File: funnel/views/proposal.py [hunk 7]
```diff
@@ -386,21 +390,24 @@ def proposal_prev(profile, space, proposal):
...
-        db.session.commit()  # WARNING: GET request!
+    proposal_move_form = ProposalMoveForm()
+    if not proposal_move_form.validate_on_submit():
+        flash(_("Please choose a proposal space you want to move this proposal to."))
         return redirect(proposal.url_for(), 303)
-    else:
-        abort(404)
+
+    target_space = proposal_move_form.target.data
+    if not (target_space.profile.admin_team in g.user.teams or target_space.admin_team in g.user.teams):
```
Comment (shreyas-satish) [inline funnel/views/proposal.py]: Is this really required? Doesn't the permission check for `move-proposal` on L398 suffice?

Comment (iambibhas) [reply funnel/views/proposal.py]: @shreyas-satish does L398 permission check if the user has permission to move proposal to the target profile?

Comment (iambibhas) [reply funnel/views/proposal.py]: I think the permission in L398 is assigned based on current user and current proposal space, right?

Comment (shreyas-satish) [reply funnel/views/proposal.py]: > I think the permission in L398 is assigned based on current user and current proposal space, right?

@iambibhas Try it out?

Comment (iambibhas) [reply funnel/views/proposal.py]: @shreyas-satish L398 says that you have `move-proposal` permission for current proposal space. We need to check explicitly if the user has same permission for the space we want to move to. Hence this check.

**Pull request merged**